util/src/main/java/com/google/cloud/hadoop/util/AbstractGoogleAsyncWriteChannel.java

```diff
@@ -128,6 +145,20 @@ public synchronized int write(ByteBuffer buffer) throws IOException {
     }
   }
 
+  private long hash(ByteBuffer src, long written) {
+    ByteBuffer buffer = src.slice();
+    int remaining = buffer.remaining();
+    int consumed = remaining;
+    if (written < remaining) {
+      int intExact = Math.toIntExact(written);
+      buffer.limit(intExact);
+      consumed = intExact;
+    }
+    cumulativeCrc32c.putBytes(buffer);
```

Comment:
It's a bit complex for me to follow, 
Shouldn't we be just be adjusting the duplicate buffer's limit to pos()+writtenBytes and provide it to the hasher to read?